// edge-tts-client/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VOLUME = exports.RATE = exports.PITCH = exports.OUTPUT_FORMAT = void 0;
var OUTPUT_FORMAT;
(function (OUTPUT_FORMAT) {
    OUTPUT_FORMAT["AUDIO_24KHZ_48KBITRATE_MONO_MP3"] = "audio-24khz-48kbitrate-mono-mp3";
    OUTPUT_FORMAT["AUDIO_24KHZ_96KBITRATE_MONO_MP3"] = "audio-24khz-96kbitrate-mono-mp3";
    OUTPUT_FORMAT["WEBM_24KHZ_16BIT_MONO_OPUS"] = "webm-24khz-16bit-mono-opus";
})(OUTPUT_FORMAT || (exports.OUTPUT_FORMAT = OUTPUT_FORMAT = {}));
var PITCH;
(function (PITCH) {
    PITCH["X_LOW"] = "x-low";
    PITCH["LOW"] = "low";
    PITCH["MEDIUM"] = "medium";
    PITCH["HIGH"] = "high";
    PITCH["X_HIGH"] = "x-high";
    PITCH["DEFAULT"] = "default";
})(PITCH || (exports.PITCH = PITCH = {}));
var RATE;
(function (RATE) {
    RATE["X_SLOW"] = "x-slow";
    RATE["SLOW"] = "slow";
    RATE["MEDIUM"] = "medium";
    RATE["FAST"] = "fast";
    RATE["X_FAST"] = "x-fast";
    RATE["DEFAULT"] = "default";
})(RATE || (exports.RATE = RATE = {}));
var VOLUME;
(function (VOLUME) {
    VOLUME["SILENT"] = "silent";
    VOLUME["X_SOFT"] = "x-soft";
    VOLUME["SOFT"] = "soft";
    VOLUME["MEDIUM"] = "medium";
    VOLUME["LOUD"] = "loud";
    VOLUME["X_LOUD"] = "x-LOUD";
    VOLUME["DEFAULT"] = "default";
})(VOLUME || (exports.VOLUME = VOLUME = {}));
