// edge-tts-client/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VOLUME = exports.RATE = exports.PITCH = exports.OUTPUT_FORMAT = exports.ProsodyOptions = exports.EdgeTTSClient = void 0;
// Re-exporting the main classes
var edge_tts_client_1 = require("./edge-tts-client");
Object.defineProperty(exports, "EdgeTTSClient", { enumerable: true, get: function () { return edge_tts_client_1.EdgeTTSClient; } });
var edge_tts_client_2 = require("./edge-tts-client");
Object.defineProperty(exports, "ProsodyOptions", { enumerable: true, get: function () { return edge_tts_client_2.ProsodyOptions; } });
// Re-exporting enums from constants.ts
var constants_1 = require("./constants");
Object.defineProperty(exports, "OUTPUT_FORMAT", { enumerable: true, get: function () { return constants_1.OUTPUT_FORMAT; } });
Object.defineProperty(exports, "PITCH", { enumerable: true, get: function () { return constants_1.PITCH; } });
Object.defineProperty(exports, "RATE", { enumerable: true, get: function () { return constants_1.RATE; } });
Object.defineProperty(exports, "VOLUME", { enumerable: true, get: function () { return constants_1.VOLUME; } });
